parse FontDescriptor and use FontBBox for y-coordinates of glyphs

lueck opened the pull request:
- This fixes #6 

- Added FontDescriptor record and a parser for it.

- Added a test for a predicate function that evaluates the Flags in FontDescriptor.

- Use FontBBox from FontDescriptor to determine the y-coordinates of glyphs instead of hard-coded 0 and 1 in Pdf.Content.FontInfo.

- Definition of Rectangle from Pdf.Document.Types was moved to Pdf.Core.Types in order to make it reusable in core and content. Since rectangles are described in the section about common data structures of the specs next to arrays, strings and numbers, core is a good place for it.

Yuras commented on document/lib/Pdf/Document/FontDict.hs:
```diff
@@ -222,3 +231,85 @@ loadUnicodeCMap pdf fontDict =
+  where
+    required = requiredInDict "FontDescriptor"
+    optional = optionalInDict "FontDescriptor"
+    nameValue' = fmap (decodeUtf8 . Name.toByteString) . nameValue
```
I think names don't have an encoding, probably it's better to keep `FontName` and `FontStretch` as `ByteString`.

lueck replied:
Yes, I agree. I've changed it.

Yuras commented on document/lib/Pdf/Document/FontDict.hs:
```diff
@@ -222,3 +231,85 @@ loadUnicodeCMap pdf fontDict =
+            lookupObject pdf ref
+
+      fontName <- required "FontName" nameValue' fd
+      fontFamily <- optional "FontFamily" (fmap decodeUtf8 . stringValue) fd
```
`FontFamily` is a byte string, it doesn't have any encoding, see section 7.9.2.4 of the specification. Lets keep it as a `ByteString`.

lueck replied:
Yes, again I've changed it.

Yuras commented on document/lib/Pdf/Document/FontDict.hs:
```diff
@@ -222,3 +231,85 @@ loadUnicodeCMap pdf fontDict =
+  fmap (sure . (`notice` (context ++ ": " ++ msg ++ " type failure")) . typeFun) .
+  HashMap.lookup key
+  where
+    msg = Text.unpack $ decodeUtf8 $ Name.toByteString key
```
`key` is a `Name`, and it may not be UTF-8 encoded. And `decodeUtf8` may throw an exception in such case.
It's OK to try to interpret it as UTF-8 encoded string, but I'd use `decodeUtf8With ignore` to make sure we never throw an exception here. The same for the `requiredInDict`.

lueck replied:
Very good point! I've replaced `decodeUtf8` with `decodeUtf8With ignore` as you suggest.

Yuras commented on content/test/test.hs:
```diff
@@ -7,10 +7,12 @@ where
 main = hspec $ do
   Test.UnicodeCMap.spec
   Test.Parser.spec
+  Test.FontDescriptor.spec
```
👍

lueck commented:
Sorry for the delay.
I've made the warnings disappear, changed the types of `FontName`, `FontFamily` and `FontStretch` to `ByteString` and fixed `optionalInDict` and `requiredInDict`.

lueck commented:
I just fixed parsing of `CharSet`. It's an ASCII string or a byte string, but no name. I found the issue when testing on TeX-generated PDFs.

Yuras commented:
Thanks! I'll check it again and merge it, hopefully tomorrow. (I'm a bit busy and might forget, so please ping me if you won't see any reply from me in the next few days. Sorry in advance.)

Yuras commented:
Looks great, thank you.
